src/twisted/conch/checkers.py

```diff
@@ -47,34 +44,115 @@
 _log = Logger()
 
 
-def _pwdGetByName(username):
+class UserRecord(Tuple[str, str, int, int, str, str, str]):
+    """
+    A record in a UNIX-style password database. See L{pwd} for field details.
+
+    This corresponds to the undocumented type L{pwd.struct_passwd}, but lacks named
+    field accessors.
+    """
+
+    @property
+    def pw_dir(self) -> str:
+        ...
+
+
+class UserDB(Protocol):
+    """
+    A database of users by name, like the stdlib L{pwd} module.
+
+    See L{twisted.python.fakepwd} for an in-memory implementation.
+    """
+
+    def getpwnam(self, username: str) -> UserRecord:
+        """
+        Lookup a user record by name.
+
+        @raises KeyError: when no such user exists
+        """
+
+
+pwd: Optional[UserDB]
+try:
+    import pwd as _pwd
+except ImportError:
+    pwd = None
+else:
+    pwd = cast(UserDB, _pwd)
+
+
+try:
+    import spwd as _spwd
+except ImportError:
+    spwd = None
+else:
+    spwd = _spwd
+
+
+class CryptedPasswordRecord(Protocol):
+    """
+    A sequence where the item at index 1 may be a crypted password.
+
+    Both L{pwd.struct_passwd} and L{spwd.struct_spwd} conform to this protocol.
+    """
+
+    def __getitem__(self, index: Literal[1]) -> str:
+        """
+        Get the crypted password.
+        """
+
+
+def _lookupUser(userdb: UserDB, username: bytes) -> UserRecord:
+    """
+    Lookup a user by name in a L{pwd}-style database.
+
+    @param userdb: The user database.
+
+    @param username: Identifying name in bytes. This will be decoded according
+    to the filesystem encoding, as the L{pwd} module does internally.
+
+    @raises KeyError: when the user doesn't exist
+    """
+    return userdb.getpwnam(username.decode(sys.getfilesystemencoding()))
```

Comment:
This looks OK as I see a similar code in `UNIXPasswordDatabase.requestAvatarId`

I was expecting the credential username and password to be `str`.
Who is using random bytes for this?

Or allow the whole Twisted cred framework to be setup with a custom encoding.

--------

I am not using the Twisted Cred framework myself.
I think that when I tried to use it 10 years ago, when I was just learning Twisted I ended up with lots of issued due to username encodings.
I had automated tests running for various encoding and at that time I didn't understood well how default encoding works and how to mock it.
...so I ended up with my custom credential code.

Reply:
The code it's actually inverting is this:

https://github.com/python/cpython/blob/a5451c96a14ac0c3ee7cef7b5c52ab1d783ec613/Modules/pwdmodule.c#L222-L226

As I grumbled on the PyPy bug I think it's weird to force us to deal in Unicode when the actual API is `bytes` with an undefined encoding, but :man_shrugging:.

Reply:
Yes. I guess that in cpython3 due to the impulse to have everything as unicode, we ended up with this API.

This is life :)